### PR TITLE
FiberSensor supports Parquet Cache

### LINKED ISSUE #702: Executor lost cause by FiberCacheManager.status method
FiberCacheManager.status assume  `dataFile` is InstanceOf `OapDataFile`, but it may be a `ParquetDataFile` instance, when dataFile is a `ParquetDataFile` instance, there will throw `ClassCastException`,  heartbeat will report failure and executor will lost cause by heartbeat timeout.

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

Fix #702 
This PR enables ParquetDataFileHandle support in FiberCacheStatus.

## How was this patch tested?

Current UT

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 1]
```diff
@@ -158,11 +158,11 @@ object FiberCacheManager extends Logging {
...
     val statusRawData = dataFibers.groupBy(_.file).map {
       case (dataFile, fiberSet) =>
-        val fileMeta = DataFileHandleCacheManager(dataFile).asInstanceOf[OapDataFileHandle]
-        val fiberBitSet = new BitSet(fileMeta.groupCount * fileMeta.fieldCount)
+        val fileMeta: DataFileHandle = DataFileHandleCacheManager(dataFile)
+        val fiberBitSet = new BitSet(fileMeta.getGroupCount * fileMeta.getFieldCount)
         fiberSet.foreach(fiber =>
-          fiberBitSet.set(fiber.columnIndex + fileMeta.fieldCount * fiber.rowGroupId))
-        FiberCacheStatus(dataFile.path, fiberBitSet, fileMeta)
+          fiberBitSet.set(fiber.columnIndex + fileMeta.getFieldCount * fiber.rowGroupId))
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: This bit set look like:
```
           field#0      field#1      field#2
group#0        0            1            0          =>  1 + 3 * 0 = 1
group#1        0            1            0          =>  1 + 3 * 1 = 4
group#2        0            0            1          =>  2 + 3 * 2 = 8
Bitset: 010010001
```

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala [hunk 2]
```diff
@@ -158,11 +158,11 @@ object FiberCacheManager extends Logging {
 
     val statusRawData = dataFibers.groupBy(_.file).map {
       case (dataFile, fiberSet) =>
-        val fileMeta = DataFileHandleCacheManager(dataFile).asInstanceOf[OapDataFileHandle]
-        val fiberBitSet = new BitSet(fileMeta.groupCount * fileMeta.fieldCount)
+        val fileMeta: DataFileHandle = DataFileHandleCacheManager(dataFile)
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: necessary. `DataFileHandleCacheManager.apply()` returns type `Nothing` if no explicit type.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFileHandle.scala [hunk 3]
```diff
@@ -35,4 +35,9 @@ private[oap] class ParquetDataFileHandle(
     this.footer = ParquetFileReader.readFooter(conf, path, NO_FILTER)
     this
   }
+
+  override def getGroupCount: Int = if (footer == null) 0 else footer.getBlocks.size()
+
+  override def getFieldCount: Int =
+    if (footer == null) 0 else footer.getFileMetaData.getSchema.getColumns.size()
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFileHandle.scala]: Perhaps we should refactor constructor of ParquetDataFileHandle, pass path & configuration and init footer in constructor, a ParquetDataFileHandle instance without initialization of footer should not exists, I think it's a bad design in the past, after this change footer no longer need to be judged empty

Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFileHandle.scala]: And for cached ParquetDataFileHandle , footer of parquet should already read in mem, no negative effect of performance

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: `Nothing ` ? why not` [T <: DataFileHandle]`?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFileHandle.scala (hunk 3 above)
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFileHandle.scala]: sure. should we separate this to next PR?

Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFileHandle.scala]: ok~

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: don't know.

[the remark this answers is not in the document]
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: may be declare as `DataFileHandleCacheManager.apply(fiberCache: DataFile): DataFileHandle`?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: Thanks, this is quite clear, how about adding a comment here? :)

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCacheManager.scala]: @linhong-intel @LuciferYang Yes, so we should merge #705 : )

**Comment by linhong-intel** [comment]
merge it. and @LuciferYang  can you create an issue for your suggestion?

> Perhaps we should refactor constructor of ParquetDataFileHandle, pass path & configuration and init footer in constructor, a ParquetDataFileHandle instance without initialization of footer should not exists, I think it's a bad design in the past, after this change footer no longer need to be judged empty

**Comment by LuciferYang** [comment]
thx~ @linhong-intel @adrian-wang

**Pull request merged**